google/cloud/storage/internal/request_parameters.h

```diff
@@ -0,0 +1,121 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_REQUEST_PARAMETERS_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_REQUEST_PARAMETERS_H_
+
+#include "google/cloud/storage/version.h"
+#include <utility>
+
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+/**
+ * Refactor common functions that manipulate list of parameters in a request.
+ *
+ * This class is used in the implementation of `RequestParameters` see below for
+ * more details.
+ */
+template <typename Parameter, typename... Parameters>
+class RequestParameterList : public RequestParameterList<Parameters...> {
+ public:
+  using RequestParameterList<Parameters...>::set_parameter;
+
+  void set_parameter(Parameter&& p) { parameter_ = std::move(p); }
+
+  template <typename HttpRequest>
+  void AddToHttpRequest(HttpRequest& request) const {
+    request.AddWellKnownParameter(parameter_);
+    RequestParameterList<Parameters...>::AddToHttpRequest(request);
+  }
+
+ private:
+  Parameter parameter_;
```

Comment:
Q: How does this work? Is there an individual private member (singular, not a parameter pack) for each `Parameter`?  Or is there one a member of type `Parameter`, and the underlying base class's parameter of type `Parameters...`?

Reply:
Ohhhh.  Thanks for the explanation!